src/mechanics/mesh/MeshGmsh.cpp

```diff
@@ -0,0 +1,378 @@
+#include "MeshGmsh.h"
+
+#include "base/Exception.h"
+#include "mechanics/interpolation/InterpolationTrussLinear.h"
+#include "mechanics/interpolation/InterpolationTriangleLinear.h"
+#include "mechanics/interpolation/InterpolationQuadLinear.h"
+
+#include <array>
+#include <climits>
+#include <algorithm>
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+
+// Helper structs (cpp only)
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+struct GmshHeader
+{
+    double version;
+    bool isBinary;
+    int doubleSize;
+};
+
+struct GmshNode
+{
+    int id = 0;
+    std::array<double, 3> coordinates;
+};
+
+struct GmshElement
+{
+    int id = 0;
+    int type = 0;
+    std::vector<int> tags;
+    std::vector<int> nodes;
+};
+
+struct GmshPhysicalNames
+{
+    int id = 0;
+    int dimension = 0;
+    std::string name;
+};
+
+
+struct GmshFileContent
+{
+    GmshHeader header;
+    int minNodeId = INT_MAX;
+    int maxNodeId = INT_MIN;
+    int dimension = 1;
+    std::vector<GmshNode> nodes;
+    std::vector<GmshElement> elements;
+    std::vector<GmshPhysicalNames> physicalNames;
+};
+
+// Helper functions (cpp only)
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+static GmshHeader ReadGmshHeader(std::ifstream& file)
+{
+    std::string line;
+    getline(file, line);
+
+    int binary;
+    GmshHeader header;
+    file >> header.version;
+    file >> binary;
+    file >> header.doubleSize;
+
+    header.isBinary = (binary == 1);
+
+    std::getline(file, line); // endl
+    if (header.version < 2.)
+    {
+        throw NuTo::Exception(__PRETTY_FUNCTION__, "Gmsh version 2.0 or higher requiered. - File version is " +
+                                                           std::to_string(header.version));
+    }
+    getline(file, line); // $EndMeshFormat
+    return header;
+}
+
+
+void ReadNodesASCII(std::istream& file, GmshFileContent& fileContent)
+{
+    std::string line;
+    std::getline(file, line);
+    int numNodes = std::stoi(line);
+
+    fileContent.nodes.resize(numNodes);
+    bool is2d = false;
+    bool is3d = false;
+    for (GmshNode& node : fileContent.nodes)
+    {
+        file >> node.id;
+        file >> node.coordinates[0];
+        file >> node.coordinates[1];
+        file >> node.coordinates[2];
+        fileContent.minNodeId = std::min(fileContent.minNodeId, node.id);
+        fileContent.maxNodeId = std::max(fileContent.maxNodeId, node.id);
+        is2d = (is2d || node.coordinates[1] != 0);
+        is3d = (is3d || node.coordinates[2] != 0);
+        std::getline(file, line);
+    }
+    if (is2d)
+        fileContent.dimension = 2;
+    if (is3d)
+        fileContent.dimension = 3;
+
+    std::getline(file, line);
```

Comment:
The following lines are duplicated in the methods below. Please refactor :)

Reply:
Not sure if I got you right, ...but seriously?

You want to introduce an own function with two parameters (file and physical name) for those four lines
~~~cpp
    std::getline(file, line);
    std::transform(line.begin(), line.end(), line.begin(), ::toupper);
    if (line.compare("$ENDPHYSICALNAMES") != 0)
        throw NuTo::Exception(__PRETTY_FUNCTION__, "$EndPhysicalNames not found!");
~~~

that are duplicated in 3 funtions that are never used outside the cpp file? I think it is a little bit too much, to be honest. While the compiler doesn't care and inlines it anyways, it is sometimes a big PITA to read a function if you have to jump from function to function in order to understand what is actually happening. If the majority wants that, I will do it, but if nobody else cares, I wont.

Reply:
Code duplication is not about how many arguments or how many lines you save. Its about making it more maintainable and readable. Imagine this likes would be
~~~cpp
ReadNextExpectedLine(file, "$EndPhysicalNames");
~~~
I would prefer this. Even if it only occurs once.

Reply:
But for the readability it is not always beneficial to split the code in multiply functions for the reason I wrote above. At this point here, I think it is more about personal  preferences than any real benefit. YMMV